lib/pull_image_to_stream.js

```diff
@@ -14,17 +16,48 @@ export const IMAGE_ERROR = 'Pulling docker image "%s" has failed this may indica
                   'Error with the registry used or an authentication error ' +
                   'in the worker try pulling the image locally. \n Error %s';
 
+export async function pullImageStreamTo(docker, image, stream, options) {
+  // Settings for exponential backoff for retrying image pulls.
+  // Last attempt will be in a range between 6 and 10 minutes which is acceptable
+  // for an image pull.
+  let maxAttempts = 5;
+  let attempts = 0;
+  let delayFactor = 15 * 1000;
+  let randomizationFactor = 0.25;
+  let error;
 
-export function pullImageStreamTo(docker, image, stream, options) {
-  return new Promise(function(accept, reject) {
-    debug('pull image', image);
-    var downloadProgress =
+  while (attempts++ < maxAttempts) {
+    error = undefined;
+
+    debug(`pull image. Image: ${image} Attempts: ${attempts}`);
+    let downloadProgress =
       dockerUtils.pullImageIfMissing(docker, image, options);
 
-    downloadProgress.pipe(stream, { end: false });
-    downloadProgress.once('error', reject);
-    downloadProgress.once('end', accept);
-  });
+    downloadProgress.pipe(stream, {end: false});
+    downloadProgress.once('error', err => { error = err; });
+    await Promise.race([
+      waitForEvent(downloadProgress, 'error'),
+      waitForEvent(downloadProgress, 'end')
+    ]);
+
+    if (!error) {
+      return;
+    }
+
+    if (attempts === maxAttempts) {
+      throw new Error(error);
+    }
+
+    let delay = Math.pow(2, attempts - 1) * delayFactor;
+    delay = delay * (Math.random() * 2 * randomizationFactor + 1 - randomizationFactor);
+    debug(
+      `pull image failed ` +
+      `Next Attempt in: ${delay} ms` +
+      `Image: ${image} ` +
+      `Error: ${error} `
+    );
```

Comment:
Don't use `debug` like this... it's much cheaper to do: `debug("pull image failed Next Attempt %s ... image: %s", delay, image, ...)` Also it's a good idea to dump `err.stack` along with `err` using `%j` so you get any properties set on error.

Example,
```js
debug('pull image failed next attempt in %s ms, Image: %s, err: %s, as JSON: %j', delay, image, err, err, err.stack);
```

The last `err.stack` will be added to it as string... this works well for me...